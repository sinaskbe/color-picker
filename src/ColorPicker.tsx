import type { CSSProperties } from 'react';
import React, { forwardRef, useMemo } from 'react';
import { ColorPickerPrefixCls, defaultColor, generateColor } from './util';

import classNames from 'classnames';
import ColorBlock from './components/ColorBlock';
import Picker from './components/Picker';
import Slider from './components/Slider';
import useColorState from './hooks/useColorState';
import type { BaseColorPickerProps, ColorGenInput } from './interface';

const hueColor = [
  'rgb(255, 0, 0) 0%',
  'rgb(255, 255, 0) 17%',
  'rgb(0, 255, 0) 33%',
  'rgb(0, 255, 255) 50%',
  'rgb(0, 0, 255) 67%',
  'rgb(255, 0, 255) 83%',
  'rgb(255, 0, 0) 100%',
];

export interface ColorPickerProps extends BaseColorPickerProps {
  value?: ColorGenInput;
  defaultValue?: ColorGenInput;
  className?: string;
  style?: CSSProperties;
  /** Get panel element  */
  panelRender?: (panel: React.ReactElement) => React.ReactElement;
  /** Disabled alpha selection */
  disabledAlpha?: boolean;
}

export default forwardRef<HTMLDivElement, ColorPickerProps>((props, ref) => {
  const {
    value,
    defaultValue,
    prefixCls = ColorPickerPrefixCls,
    onChange,
    onChangeComplete,
    className,
    style,
    panelRender,
    disabledAlpha = false,
    disabled = false,
  } = props;
  const [colorValue, setColorValue] = useColorState(defaultColor, {
    value,
    defaultValue,
  });
  const alphaColor = useMemo(() => {
    const rgb = generateColor(colorValue.toRgbString());
    // alpha color need equal 1 for base color
    rgb.setAlpha(1);
    return rgb.toRgbString();
  }, [colorValue]);
  const mergeCls = classNames(`${prefixCls}-panel`, className, {
    [`${prefixCls}-panel-disabled`]: disabled,
  });
  const basicProps = {
    prefixCls,
    onChangeComplete,
    disabled,
  };

  const handleChange: BaseColorPickerProps['onChange'] = (data, type) => {
    if (!value) {
      setColorValue(data);
    }
    onChange?.(data, type);
  };

  const defaultPanel = (
    <>
      <Picker color={colorValue} onChange={handleChange} {...basicProps} />
      <div className={`${prefixCls}-slider-container`}>
        <div
          className={classNames(`${prefixCls}-slider-group`, {
            [`${prefixCls}-slider-group-disabled-alpha`]: disabledAlpha,
          })}
        >
          <Slider
            gradientColors={hueColor}
            color={colorValue}
            value={`hsl(${colorValue.toHsb().h},100%, 50%)`}
            onChange={color => handleChange(color, 'hue')}
            {...basicProps}
          />
          {!disabledAlpha && (
            <Slider
              type="alpha"
              gradientColors={['rgba(255, 0, 4, 0) 0%', alphaColor]}
              color={colorValue}
              value={colorValue.toRgbString()}
              onChange={color => handleChange(color, 'alpha')}
              {...basicProps}
            />
          )}
        </div>
        <ColorBlock color={colorValue.toRgbString()} prefixCls={prefixCls} />
      </div>
    </>
  );

  return (
    <div className={mergeCls} style={style} ref={ref}>
      {typeof panelRender === 'function'
        ? panelRender(defaultPanel)
        : defaultPanel}
    </div>
  );
});
