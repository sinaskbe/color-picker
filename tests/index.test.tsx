/* eslint-disable @typescript-eslint/no-loop-func */
import { act, createEvent, fireEvent, render } from '@testing-library/react';
import { spyElementPrototypes } from 'rc-util/lib/test/domHook';
import React, { useState } from 'react';
import ColorPicker from '../src';
import { defaultColor } from '../src/util';

vi.mock('@rc-component/trigger', async () => {
  return await import('@rc-component/trigger/lib/mock');
});

export async function waitFakeTimer(advanceTime = 1000, times = 20) {
  for (let i = 0; i < times; i += 1) {
    // eslint-disable-next-line no-await-in-loop
    await act(async () => {
      await Promise.resolve();

      if (advanceTime > 0) {
        vi.advanceTimersByTime(advanceTime);
      } else {
        vi.runAllTimers();
      }
    });
  }
}

function doMouseMove(
  container,
  start,
  end,
  element = 'rc-color-picker-handler',
) {
  const mouseDown = createEvent.mouseDown(
    container.getElementsByClassName(element)[0],
    {
      pageX: start,
      pageY: start,
    },
  );
  fireEvent(container.getElementsByClassName(element)[0], mouseDown);

  // Drag
  const mouseMove: any = new Event('mousemove');
  mouseMove.pageX = end;
  mouseMove.pageY = end;

  fireEvent(document, mouseMove);

  const mouseUp = createEvent.mouseUp(document);
  fireEvent(document, mouseUp);
}

function doTouchMove(
  container,
  start,
  end,
  element = 'rc-color-picker-handler',
) {
  const touchStart: any = createEvent.touchStart(
    container.getElementsByClassName(element)[0],
    {
      touches: [{}],
    },
  );
  touchStart.touches[0].pageX = start;
  fireEvent(container.getElementsByClassName(element)[0], touchStart);

  // Drag
  const touchMove: any = createEvent.touchMove(document, {
    touches: [{}],
  });
  touchMove.touches[0].pageX = end;
  fireEvent(document, touchMove);
}

describe('ColorPicker', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });
  afterEach(() => {
    vi.useRealTimers();
  });

  it('Should component render correct', () => {
    const { container } = render(<ColorPicker defaultValue={defaultColor} />);
    expect(container).toMatchSnapshot();
    expect(container.querySelector('.rc-color-picker-panel')).toBeTruthy();
  });

  it('Should component onChange work on no control mode', () => {
    spyElementPrototypes(HTMLElement, {
      getBoundingClientRect: () => ({
        x: 0,
        y: 100,
        width: 100,
        height: 100,
      }),
    });
    const handleChange = vi.fn();
    const { container } = render(<ColorPicker onChange={handleChange} />);
    expect(container).toMatchSnapshot();
    expect(container.querySelector('.rc-color-picker-panel')).toBeTruthy();
    doMouseMove(container, 0, 999);
    expect(handleChange).toBeCalled();
  });

  it('Should pick color work by mouse', () => {
    spyElementPrototypes(HTMLElement, {
      getBoundingClientRect: () => ({
        x: 0,
        y: 100,
        width: 100,
        height: 100,
      }),
    });
    const App = () => {
      const [value, setValue] = useState(defaultColor);
      return (
        <>
          <ColorPicker value={value} onChange={setValue} />
          <div className="pick-color">{value.toHsbString()}</div>
          <div>{value.toHexString()}</div>
          <div>{value.toHex8String()}</div>
        </>
      );
    };
    const { container } = render(<App />);
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(215, 91%, 100%)',
    );

    doMouseMove(container, 0, 999);
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(215, 100%, 0%)',
    );

    doMouseMove(
      container.querySelector('.rc-color-picker-slider-hue'),
      0,
      9999,
    );
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(360, 100%, 0%)',
    );

    doMouseMove(
      container.querySelector('.rc-color-picker-slider-alpha'),
      9999,
      0,
    );
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsba(360, 100%, 0%, 0)',
    );
  });

  it('Should no control pick color work', () => {
    spyElementPrototypes(HTMLElement, {
      getBoundingClientRect: () => ({
        x: 0,
        y: 100,
        width: 100,
        height: 100,
      }),
    });
    const App = () => {
      const [value, setValue] = useState(defaultColor);
      return (
        <>
          <ColorPicker value={value} onChange={setValue} />
          <div className="pick-color">{value.toHsbString()}</div>
          <div>{value.toHexString()}</div>
          <div>{value.toHex8String()}</div>
        </>
      );
    };
    const { container } = render(<App />);
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(215, 91%, 100%)',
    );

    doMouseMove(container, 0, 9999);
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(215, 100%, 0%)',
    );

    doMouseMove(
      container.querySelector('.rc-color-picker-slider-hue'),
      0,
      9999,
    );
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(360, 100%, 0%)',
    );

    doMouseMove(
      container.querySelector('.rc-color-picker-slider-alpha'),
      9999,
      0,
    );
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsba(360, 100%, 0%, 0)',
    );
  });

  it('Should pick color work by touch', () => {
    spyElementPrototypes(HTMLElement, {
      getBoundingClientRect: () => ({
        x: 0,
        y: 100,
        width: 100,
        height: 100,
      }),
    });
    const App = () => {
      const [value, setValue] = useState(defaultColor);
      return (
        <>
          <div className="pick-color">{value.toHsbString()}</div>
          <ColorPicker value={value} onChange={setValue} />
        </>
      );
    };
    const { container } = render(<App />);
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(215, 91%, 100%)',
    );

    doTouchMove(container, 0, 9999);
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(215, 100%, 0%)',
    );

    doTouchMove(
      container.querySelector('.rc-color-picker-slider-hue'),
      0,
      9999,
    );
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(360, 100%, 0%)',
    );

    doTouchMove(
      container.querySelector('.rc-color-picker-slider-alpha'),
      9999,
      0,
    );
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsba(360, 100%, 0%, 0)',
    );
  });

  it('Should custom panel work', () => {
    const { container } = render(
      <ColorPicker
        panelRender={panel => <div className="custom-panel">{panel}</div>}
      />,
    );
    expect(container.querySelector('.custom-panel')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('Should prefixCls work', () => {
    const { container } = render(<ColorPicker prefixCls="test-prefixCls" />);
    expect(container).toMatchSnapshot();
  });

  it('Should drag boundary cases work', () => {
    spyElementPrototypes(HTMLElement, {
      getBoundingClientRect: () => ({
        x: 0,
        y: 100,
        width: 0,
        height: 0,
      }),
    });
    const App = () => {
      const [value, setValue] = useState(defaultColor);
      return (
        <>
          <div className="pick-color">{value.toHsbString()}</div>
          <ColorPicker value={value} onChange={setValue} />
        </>
      );
    };
    const { container } = render(<App />);
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(215, 91%, 100%)',
    );
    doMouseMove(container, 0, 9999);
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(215, 91%, 100%)',
    );
  });

  it('Should hsb string work', () => {
    const App = () => <ColorPicker value={'hsb(215, 91%, 100%)'} />;
    const { container } = render(<App />);
    expect(
      container.querySelector('.rc-color-picker-handler').getAttribute('style'),
    ).toEqual('background-color: rgb(23, 120, 255);');
  });

  it('Should rgb string work', () => {
    const App = () => <ColorPicker value={'rgb(23, 120, 255)'} />;
    const { container } = render(<App />);
    expect(
      container.querySelector('.rc-color-picker-handler').getAttribute('style'),
    ).toEqual('background-color: rgb(23, 120, 255);');
  });

  it('Should hex string work', () => {
    const App = () => <ColorPicker value="#1778ff" />;
    const { container } = render(<App />);
    expect(
      container.querySelector('.rc-color-picker-handler').getAttribute('style'),
    ).toEqual('background-color: rgb(23, 120, 255);');
  });

  it('Should hsb obj work', () => {
    const App = () => <ColorPicker value={{ h: 215, s: 0.91, b: 1 }} />;
    const { container } = render(<App />);
    expect(
      container.querySelector('.rc-color-picker-handler').getAttribute('style'),
    ).toEqual('background-color: rgb(23, 120, 255);');
  });

  it('Should rgb obj work', () => {
    const App = () => <ColorPicker value={{ r: 23, g: 120, b: 255 }} />;
    const { container } = render(<App />);
    expect(
      container.querySelector('.rc-color-picker-handler').getAttribute('style'),
    ).toEqual('background-color: rgb(23, 120, 255);');
  });

  it('Should disabled work', () => {
    spyElementPrototypes(HTMLElement, {
      getBoundingClientRect: () => ({
        x: 0,
        y: 100,
        width: 0,
        height: 0,
      }),
    });
    const handleChange = vi.fn();
    const App = () => <ColorPicker disabled onChange={handleChange} />;
    const { container } = render(<App />);
    expect(
      container.querySelector('.rc-color-picker-panel-disabled'),
    ).toBeTruthy();
    doMouseMove(container, 0, 9999);
    expect(handleChange).toBeCalledTimes(0);
  });

  it('Should disabled alpha work', () => {
    spyElementPrototypes(HTMLElement, {
      getBoundingClientRect: () => ({
        x: 0,
        y: 100,
        width: 0,
        height: 0,
      }),
    });
    const App = () => <ColorPicker disabledAlpha />;
    const { container } = render(<App />);
    expect(
      container.querySelector('.rc-color-picker-slider-alpha'),
    ).toBeFalsy();
    expect(container).toMatchSnapshot();
  });

  it('Should onChangeComplete work', () => {
    const handleChange = vi.fn();
    const App = () => <ColorPicker onChangeComplete={handleChange} />;
    const { container } = render(<App />);
    doMouseMove(container, 0, 9999);
    doMouseMove(
      container.querySelector('.rc-color-picker-slider-alpha'),
      0,
      9999,
    );
    doMouseMove(
      container.querySelector('.rc-color-picker-slider-hue'),
      0,
      9999,
    );
    expect(handleChange).toBeCalledTimes(3);
    expect(handleChange).toHaveBeenLastCalledWith('hue');
  });
});
