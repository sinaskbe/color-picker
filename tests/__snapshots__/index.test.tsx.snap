// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ColorPicker > Should component onChange work on no control mode 1`] = `
<div>
  <div
    class="rc-color-picker-panel"
  >
    <div
      class="rc-color-picker-select"
    >
      <div
        class="rc-color-picker-palette"
        style="position: relative;"
      >
        <div
          style="position: absolute; left: 41.372549019607845px; top: -50px; z-index: 1;"
        >
          <div
            class="rc-color-picker-handler"
            style="background-color: rgb(22, 119, 255);"
          />
        </div>
        <div
          class="rc-color-picker-saturation"
          style="background-color: rgb(0, 0, 0);"
        />
      </div>
    </div>
    <div
      class="rc-color-picker-slider-container"
    >
      <div
        class="rc-color-picker-slider-group"
      >
        <div
          class="rc-color-picker-slider rc-color-picker-slider-hue"
        >
          <div
            class="rc-color-picker-palette"
            style="position: relative;"
          >
            <div
              style="position: absolute; left: 9.728183118741065px; top: -16.666666666666668px; z-index: 1;"
            >
              <div
                class="rc-color-picker-handler rc-color-picker-handler-sm"
                style="background-color: rgb(0, 0, 0);"
              />
            </div>
            <div
              class="rc-color-picker-gradient"
              style="position: absolute; inset: 0;"
            />
          </div>
        </div>
        <div
          class="rc-color-picker-slider rc-color-picker-slider-alpha"
        >
          <div
            class="rc-color-picker-palette"
            style="position: relative;"
          >
            <div
              style="position: absolute; left: 50px; top: -16.666666666666668px; z-index: 1;"
            >
              <div
                class="rc-color-picker-handler rc-color-picker-handler-sm"
                style="background-color: rgb(22, 119, 255);"
              />
            </div>
            <div
              class="rc-color-picker-gradient"
              style="position: absolute; inset: 0;"
            />
          </div>
        </div>
      </div>
      <div
        class="rc-color-picker-color-block"
      >
        <div
          class="rc-color-picker-color-block-inner"
          style="background: rgb(22, 119, 255);"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ColorPicker > Should component render correct 1`] = `
<div>
  <div
    class="rc-color-picker-panel"
  >
    <div
      class="rc-color-picker-select"
    >
      <div
        class="rc-color-picker-palette"
        style="position: relative;"
      >
        <div
          style="position: absolute; left: 0px; top: 0px; z-index: 1;"
        >
          <div
            class="rc-color-picker-handler"
            style="background-color: rgb(22, 119, 255);"
          />
        </div>
        <div
          class="rc-color-picker-saturation"
          style="background-color: rgb(0, 0, 0);"
        />
      </div>
    </div>
    <div
      class="rc-color-picker-slider-container"
    >
      <div
        class="rc-color-picker-slider-group"
      >
        <div
          class="rc-color-picker-slider rc-color-picker-slider-hue"
        >
          <div
            class="rc-color-picker-palette"
            style="position: relative;"
          >
            <div
              style="position: absolute; left: 0px; top: 0px; z-index: 1;"
            >
              <div
                class="rc-color-picker-handler rc-color-picker-handler-sm"
                style="background-color: rgb(0, 0, 0);"
              />
            </div>
            <div
              class="rc-color-picker-gradient"
              style="position: absolute; inset: 0;"
            />
          </div>
        </div>
        <div
          class="rc-color-picker-slider rc-color-picker-slider-alpha"
        >
          <div
            class="rc-color-picker-palette"
            style="position: relative;"
          >
            <div
              style="position: absolute; left: 0px; top: 0px; z-index: 1;"
            >
              <div
                class="rc-color-picker-handler rc-color-picker-handler-sm"
                style="background-color: rgb(22, 119, 255);"
              />
            </div>
            <div
              class="rc-color-picker-gradient"
              style="position: absolute; inset: 0;"
            />
          </div>
        </div>
      </div>
      <div
        class="rc-color-picker-color-block"
      >
        <div
          class="rc-color-picker-color-block-inner"
          style="background: rgb(22, 119, 255);"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ColorPicker > Should custom panel work 1`] = `
<div>
  <div
    class="rc-color-picker-panel"
  >
    <div
      class="custom-panel"
    >
      <div
        class="rc-color-picker-select"
      >
        <div
          class="rc-color-picker-palette"
          style="position: relative;"
        >
          <div
            style="position: absolute; left: 41.372549019607845px; top: -50px; z-index: 1;"
          >
            <div
              class="rc-color-picker-handler"
              style="background-color: rgb(22, 119, 255);"
            />
          </div>
          <div
            class="rc-color-picker-saturation"
            style="background-color: rgb(0, 0, 0);"
          />
        </div>
      </div>
      <div
        class="rc-color-picker-slider-container"
      >
        <div
          class="rc-color-picker-slider-group"
        >
          <div
            class="rc-color-picker-slider rc-color-picker-slider-hue"
          >
            <div
              class="rc-color-picker-palette"
              style="position: relative;"
            >
              <div
                style="position: absolute; left: 9.728183118741065px; top: -16.666666666666668px; z-index: 1;"
              >
                <div
                  class="rc-color-picker-handler rc-color-picker-handler-sm"
                  style="background-color: rgb(0, 0, 0);"
                />
              </div>
              <div
                class="rc-color-picker-gradient"
                style="position: absolute; inset: 0;"
              />
            </div>
          </div>
          <div
            class="rc-color-picker-slider rc-color-picker-slider-alpha"
          >
            <div
              class="rc-color-picker-palette"
              style="position: relative;"
            >
              <div
                style="position: absolute; left: 50px; top: -16.666666666666668px; z-index: 1;"
              >
                <div
                  class="rc-color-picker-handler rc-color-picker-handler-sm"
                  style="background-color: rgb(22, 119, 255);"
                />
              </div>
              <div
                class="rc-color-picker-gradient"
                style="position: absolute; inset: 0;"
              />
            </div>
          </div>
        </div>
        <div
          class="rc-color-picker-color-block"
        >
          <div
            class="rc-color-picker-color-block-inner"
            style="background: rgb(22, 119, 255);"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ColorPicker > Should disabled alpha work 1`] = `
<div>
  <div
    class="rc-color-picker-panel"
  >
    <div
      class="rc-color-picker-select"
    >
      <div
        class="rc-color-picker-palette"
        style="position: relative;"
      >
        <div
          style="position: absolute; left: 0px; top: 0px; z-index: 1;"
        >
          <div
            class="rc-color-picker-handler"
            style="background-color: rgb(22, 119, 255);"
          />
        </div>
        <div
          class="rc-color-picker-saturation"
          style="background-color: rgb(0, 0, 0);"
        />
      </div>
    </div>
    <div
      class="rc-color-picker-slider-container"
    >
      <div
        class="rc-color-picker-slider-group rc-color-picker-slider-group-disabled-alpha"
      >
        <div
          class="rc-color-picker-slider rc-color-picker-slider-hue"
        >
          <div
            class="rc-color-picker-palette"
            style="position: relative;"
          >
            <div
              style="position: absolute; left: 0px; top: 0px; z-index: 1;"
            >
              <div
                class="rc-color-picker-handler rc-color-picker-handler-sm"
                style="background-color: rgb(0, 0, 0);"
              />
            </div>
            <div
              class="rc-color-picker-gradient"
              style="position: absolute; inset: 0;"
            />
          </div>
        </div>
      </div>
      <div
        class="rc-color-picker-color-block"
      >
        <div
          class="rc-color-picker-color-block-inner"
          style="background: rgb(22, 119, 255);"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ColorPicker > Should prefixCls work 1`] = `
<div>
  <div
    class="test-prefixCls-panel"
  >
    <div
      class="test-prefixCls-select"
    >
      <div
        class="test-prefixCls-palette"
        style="position: relative;"
      >
        <div
          style="position: absolute; left: 41.372549019607845px; top: -50px; z-index: 1;"
        >
          <div
            class="test-prefixCls-handler"
            style="background-color: rgb(22, 119, 255);"
          />
        </div>
        <div
          class="test-prefixCls-saturation"
          style="background-color: rgb(0, 0, 0);"
        />
      </div>
    </div>
    <div
      class="test-prefixCls-slider-container"
    >
      <div
        class="test-prefixCls-slider-group"
      >
        <div
          class="test-prefixCls-slider test-prefixCls-slider-hue"
        >
          <div
            class="test-prefixCls-palette"
            style="position: relative;"
          >
            <div
              style="position: absolute; left: 9.728183118741065px; top: -16.666666666666668px; z-index: 1;"
            >
              <div
                class="test-prefixCls-handler test-prefixCls-handler-sm"
                style="background-color: rgb(0, 0, 0);"
              />
            </div>
            <div
              class="test-prefixCls-gradient"
              style="position: absolute; inset: 0;"
            />
          </div>
        </div>
        <div
          class="test-prefixCls-slider test-prefixCls-slider-alpha"
        >
          <div
            class="test-prefixCls-palette"
            style="position: relative;"
          >
            <div
              style="position: absolute; left: 50px; top: -16.666666666666668px; z-index: 1;"
            >
              <div
                class="test-prefixCls-handler test-prefixCls-handler-sm"
                style="background-color: rgb(22, 119, 255);"
              />
            </div>
            <div
              class="test-prefixCls-gradient"
              style="position: absolute; inset: 0;"
            />
          </div>
        </div>
      </div>
      <div
        class="test-prefixCls-color-block"
      >
        <div
          class="test-prefixCls-color-block-inner"
          style="background: rgb(22, 119, 255);"
        />
      </div>
    </div>
  </div>
</div>
`;
